### PR TITLE
Avoid loading dataloaders if `limit_batches=0`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes: #11587

in short: if we say `limit_val_batches=0` means no validation, why even load the dataloaders?

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: tests/models/test_hooks.py [hunk 1]
```diff
@@ -716,8 +713,7 @@ def test_trainer_model_hook_system_eval(tmpdir, batches, verb, noun, dataloader,
         dict(name="setup", kwargs=dict(stage=verb)),
         dict(name="configure_sharded_model"),
         dict(name="Callback.on_configure_sharded_model", args=(trainer, model)),
-        dict(name=f"on_{dataloader}_dataloader"),
-        dict(name=f"{dataloader}_dataloader"),
+        *([dict(name=f"on_{dataloader}_dataloader"), dict(name=f"{dataloader}_dataloader")] if batches else []),
         *(hooks if batches else []),
```
Comment (carmocca) [inline tests/models/test_hooks.py:717]: Can be merged with this line

File: tests/models/test_hooks.py [hunk 2]
```diff
@@ -643,9 +643,6 @@ def test_trainer_model_hook_system_fit_no_val_and_resume(tmpdir):
         dict(name="train", args=(True,)),
         dict(name="on_train_dataloader"),
         dict(name="train_dataloader"),
-        # even though no validation runs, we initialize the val dataloader for properties like `num_val_batches`
```
Comment (carmocca) [inline tests/models/test_hooks.py:646]: I remember noticing this and adding this comment because removing it wasn't straightforward. I'm glad we can do it now

File: tests/trainer/test_dataloaders.py [hunk 3]
```diff
@@ -234,22 +234,18 @@ def on_test_epoch_start(self, trainer, pl_module):
         self.test_epoch_count += 1
 
 
-@pytest.mark.parametrize(
-    ["limit_train_batches", "limit_val_batches", "limit_test_batches"], [(0.0, 0.0, 0.0), (1.0, 1.0, 1.0)]
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:238]: I haven't thoroughly reviewed yet but why are you removing all the tests that pass `0`?

Comment (rohitgr7) [reply tests/trainer/test_dataloaders.py:238]: yes, because I have added a test that checks that dataloader isn't loaded with 0.

**Review by tchaton — APPROVED** [review]
Very neat work !

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -2246,3 +2246,22 @@ def test_trainer_config_strategy(trainer_kwargs, expected, monkeypatch):
     assert len(expected) == 4
     for k, v in expected.items():
         assert getattr(trainer, k) == v, f"Failed {k}: {v}"
+
+
+@pytest.mark.parametrize(
+    "running_stage", [RunningStage.TRAINING, RunningStage.VALIDATING, RunningStage.TESTING, RunningStage.PREDICTING]
+)
+def test_dataloaders_are_not_loaded_if_limit_batches_is_disabled(running_stage):
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:2254]:
```suggestion
def test_dataloaders_are_not_loaded_if_disabled_through_limit_batches(running_stage):
```

File: tests/trainer/test_dataloaders.py [hunk 5]
```diff
@@ -396,24 +378,22 @@ def test_datasets_dataloaders_with_limit_num_batches(
...
-    assert trainer.num_test_batches[0] == limit_test_batches
-    assert epoch_cb.test_epoch_count == int(limit_test_batches > 0)
+    assert trainer.num_test_batches[0] == limit_batches
+    assert epoch_cb.test_epoch_count == 1
 
 
+@pytest.mark.skip()
 @pytest.mark.parametrize(
     ["limit_train_batches", "limit_val_batches", "limit_test_batches"],
-    [(0.0, 0.0, 0.0), (0, 0, 0.5), (1.0, 1.0, 1.0), (0.2, 0.4, 0.4)],
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:416]: should we group the new test together with the tests in this file? It might not be obvious that the removed cases are covered by a different test function.

File: tests/trainer/test_dataloaders.py [hunk 6]
```diff
@@ -311,37 +302,32 @@ def test_dataloaders_with_limit_train_batches(tmpdir, dataset, limit_train_batch
...
-def test_dataloaders_with_limit_val_batches(tmpdir, dataset, limit_val_batches):
+def test_dataloaders_with_limit_val_batches(tmpdir, dataset):
     """Verify inf train, val & test dataloaders (e.g. IterableDataset) passed with batch limit as number."""
 
     epoch_cb = Counter()
     callbacks = [epoch_cb]
     enable_checkpointing = False
-    if limit_val_batches > 0:
-        callbacks.append(ModelCheckpoint(monitor="val_log", save_top_k=1, mode="max", verbose=False))
-        enable_checkpointing = True
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:338]: since you are removing the case where limit_val_batches == 0, this if statement would have to stay.

**Review by Borda — APPROVED** [review]
lgtm
